python/revng/internal/cli/_commands/mass_testing/generate_report/stacktrace_transform.py

```diff
@@ -0,0 +1,99 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import dataclasses
+import json
+import os
+import sys
+from io import TextIOWrapper
+from pathlib import Path
+from subprocess import run
+from typing import Iterable, List
+
+from .test_directory import TestDirectory
+
+SCRIPT_DIR = Path(__file__).parent.resolve()
+
+
+@dataclasses.dataclass
+class StacktraceLine:
+    path: str
+    symbol: str
+
+    def to_string(self) -> str:
+        return f"{self.symbol} at {self.path}".replace(";", "")
+
+
+def normalize_path(path: str):
+    path = os.path.normpath(path)
+    # TODO: un-orchestra this
+    if "/orchestra/root/" in path:
+        return path.split("/orchestra/root/", 1)[1]
+    elif "/orchestra/sources/" in path:
+        return path.split("/orchestra/sources/", 1)[1]
+    else:
+        return path
+
+
+def llvm_stacktrace_transform_entry(entry: dict, symbol: dict) -> StacktraceLine:
+    if symbol["FunctionName"]:
+        outname = symbol["FunctionName"]
+        if symbol["StartAddress"]:
+            start_addr = int(symbol["StartAddress"], 16)
+            addr = int(entry["Address"], 16)
+            outname += f"+{addr - start_addr:#x}"
+    else:
+        outname = entry["Address"]
+
+    if symbol["FileName"]:
+        path = f"{normalize_path(symbol['FileName'])}:{symbol['Line']}"
+    else:
+        path = normalize_path(entry["ModuleName"])
+
+    return StacktraceLine(path, outname)
+
+
+def llvm_stacktrace_transform(data) -> List[StacktraceLine]:
+    results = []
+    for line in data:
+        entry = json.loads(line)
+        for symbol in entry["Symbol"]:
+            results.append(llvm_stacktrace_transform_entry(entry, symbol))
+
+    return results
+
+
+def get_perf_line(stracktrace_file: TextIOWrapper):
+    out = llvm_stacktrace_transform(stracktrace_file)
+    return ";".join(e.to_string() for e in out) + " 1\n"
+
+
+def generate_flamegraph(tests: Iterable[TestDirectory], output: str | Path, inverted: bool = False):
+    lines = ""
+    for test in tests:
+        with test.stacktrace() as f:
+            if f is not None:
+                lines += get_perf_line(f)
+            else:
+                lines += "no stack trace 1\n"
+
+    opts = ["--color=chain", "--width=1920", "--minwidth=0", "--countname=run(s)"]
+    if inverted:
+        opts.append("--inverted")
+        opts.append("--title=Stack traces - Top down - Thread entry point is at the bottom")
+    else:
+        opts.append("--title=Stack traces - Bottom up - Crash location is at the top")
```

Comment:
I'd consider having these arguments passed in from the user and forward them to `flamegraph.pl`.

For instance, this script is used with non-crashes too (while we hardcode "crash location").

Reply:
> I'd consider having these arguments passed in from the user

Do you mean `generate_report/__main__.py` or the actual user? I'd find it kind of cumbersome for `mass-testing generate-report` to provide all the command-line options.

Also, in both situations it forces the user to know what `flamegraph.pl` options are
The only thing that changes is the title, I'd rather add it as a parameter

Reply:
OK

Reply:
Done